### PR TITLE
[3/N] Core generator abstraction

### PR DESCRIPTION
**Pull request by Ritesh1905**
A split version of [this RFC PR](https://github.com/meta-pytorch/forge/pull/149)

### CODE REVIEW

File: src/forge/trainers/huggingface_trainer.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+from forge.data_models.api import Trainer, WeightsBuffer
+
+from forge.data_models.distributed_metric import Fraction, SumDistributedMetric
+
+from forge.data_models.loss import LossOutput
+from forge.data_models.minibatch import Minibatch
+from forge.stores.in_memory_store import InMemoryStore
+
+
+class HuggingFaceTrainer(Trainer):
```
Comment (felipemello1) [inline src/forge/trainers/huggingface_trainer.py:18]: At first glance, I am a bit skeptical that having multiple Trainers would work in practice. Is the idea that there would be a HF and a Titan trainer? Could you share a bit about why we need the Trainer abstraction?

Comment (Ritesh1905) [reply src/forge/trainers/huggingface_trainer.py:18]: Yes providing an optionality for the end-users. if they want to compare two runs or set bench marks or just want to try them out.

The idea is to define the abstraction and enable the policy and learner actors to inject an implementation of the trainer. Essentially what trainer/generator to use can be provided from config.

Comment (felipemello1) [reply src/forge/trainers/huggingface_trainer.py:18]: Wouldn't they have significantly different config signatures / distributed APIs / library requirements, etc? Do you see it working for a V0 of Forge?

File: src/forge/data_models/completion.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+from dataclasses import dataclass
+from typing import Optional
+
+import torch
+from forge.data_models.prompt import Prompt
+
+
+@dataclass
+class Completion:
+    """A model-generated completion for a given prompt."""
```
Comment (felipemello1) [inline src/forge/data_models/completion.py:16]: i think we already have a dataclass for this. Not sure if its "Episode". But it makes sense to have a completion class.

File: src/forge/data_models/distributed_metric.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+import torch
+import torch.distributed as dist
+
+
+@dataclass
+class DistributedMetric(ABC):
```
Comment (felipemello1) [inline src/forge/data_models/distributed_metric.py:15]: I think we should use https://github.com/meta-pytorch/forge/tree/main/src/forge/data/dataset_metrics

But lets connect and see if you think it makes sense!

File: src/forge/data_models/experience.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+from dataclasses import dataclass
+from typing import Optional, Sequence
+
+import torch
+from forge.data_models.scored_completion import ScoredCompletion
+
+
+@dataclass
+class Experience:
```
Comment (felipemello1) [inline src/forge/data_models/experience.py:15]: This is very similar to Completion. I wonder if we should just keep one. But its also a bit similar to "Episode".

Comment (felipemello1) [reply src/forge/data_models/experience.py:15]: Need to see how this plays with data packing. It could be that the Packer takes list[Experience]. But i dont think it makes sense for this class to have concat logic if we are having packing.

File: src/forge/trainers/huggingface_trainer.py (hunk 1 above)
Comment (Ritesh1905) [reply src/forge/trainers/huggingface_trainer.py:18]: Yes, they would sure have different configs and we need have a config abstraction too.

> Do you see it working for a V0 of Forge?

the current apps(grpo) use HF, and I just saw another PR to make the grpo app running for titan. so we already have two variations of it.

File: src/forge/data_models/loss.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+@dataclass
+class LossInput:
+    minibatch: Minibatch
+    trainer_logits: torch.Tensor
+
+
+@dataclass
+class LossOutput:
+    loss: Fraction
```
Comment (felipemello1) [inline src/forge/data_models/loss.py:22]: its interesting to have a LossInput and LossOutput class. I am just afraid that these abstractions would add too much hierarchy. i.e. instead of `loss(logits, targets, mask)`, we have `loss(LossInput(MiniBatch(something_else)))`.

But the LossOutput makes more sense to me because we may be outputting multiple numbers. We should double check, because if its only a couple, then having a dataclass might be an overkill

File: src/forge/trainers/huggingface_trainer.py (hunk 1 above)
Comment (felipemello1) [reply src/forge/trainers/huggingface_trainer.py:18]: @allenwang28 , do you know if we will support both Titan and HF after we work on Titan integration? At least in torchtune finding all the bugs/optimizations/features for a single framework was a lot of work.

File: src/forge/data_models/experience.py (hunk 4 above)
Comment (Ritesh1905) [reply src/forge/data_models/experience.py:15]: This is modeled in a way that, Completion is what we receive from the generator and the experience is what we feed into trainer (in-between it goes through scoring, post-processing, etc.).

File: src/forge/data_models/minibatch.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+
+def from_experiences(
+    exps: Sequence[Experience], max_seq_len: int, pad_val: int = 0
+) -> Minibatch:
+    """
+    Convert a list of experiences to a minibatch.
+    """
+
+    def pack_sequence(
```
Comment (felipemello1) [inline src/forge/data_models/minibatch.py:49]: i think we will have to revisit it after we do dataset packing. My idea is to have PackedDataset(iterator), where the iterator is a replay buffer. The replay buffer outputs an Experience, and outputs a PackedMiniBatch (or something like that)

File: src/forge/data_models/prompt.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    USER = "user"
+    ASSISTANT = "assistant"
+    NONE = "none"
+
+
+@dataclass
+class Message:
+    """A single message in a conversation."""
+
+    chunks: Sequence[str]
```
Comment (felipemello1) [inline src/forge/data_models/prompt.py:24]: whats a chunk? Is it for example 2 consecutive messages of an user?

File: src/forge/data_models/prompt.py
Comment (felipemello1) [inline src/forge/data_models/prompt.py:1]: At least in torchtune, we needed different types of messages/prompt handling. This makes sense to me for a V0, but i wonder if users will want to do more with this: https://github.com/pytorch/torchtune/blob/main/torchtune/data/_messages.py

Maybe @joecummings has some thoughts?

Comment (Ritesh1905) [reply src/forge/data_models/prompt.py:1]: You are right. I thought I left a TODO that we need support other formats like image (bytes) etc.

File: src/forge/data_models/prompt.py (hunk 7 above)
Comment (Ritesh1905) [reply src/forge/data_models/prompt.py:24]: I think the inital ideas was mark each chunk as trainable (mask). I probably complicated this by. we can just use str for now.

File: src/forge/data_models/api.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+from forge.data_models.loss import LossOutput
+from forge.data_models.minibatch import Minibatch
+from forge.data_models.prompt import Prompt
+
+
+# TODO: This file needs should NOT be in the data_models folder/package
+
+
+class Store(ABC):
```
Comment (LucasLLC) [inline src/forge/data_models/api.py:20]: This looks really close to the KVStore we're building in `https://github.com/meta-pytorch/forge/pull/147`

File: src/forge/data_models/api.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        """Store a value under a key."""
+        pass
+
+    @abstractmethod
+    def get(self, key: str) -> Any:
+        """Retrieve a value by key."""
+        pass
+
+
+class WeightsBuffer:
```
Comment (LucasLLC) [inline src/forge/data_models/api.py:52]: I think this is a reasonable abstraction in forge, although I do wonder if this is something torchstore should support to begin with? 
e.g. `(e.g., in-memory, RDMA, file system, torchstore etc.)` 

These are all things torchstore should support OOB.

File: src/forge/data_models/api.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        Retrieve the weights stored under the specified key.
+        Args:
+            key (str): The key for which to retrieve the weights.
+        Returns:
+            The weights object or reference associated with the key.
+        """
+        return self.store.get(key)
+
+
+class Trainer(ABC):
```
Comment (LucasLLC) [inline src/forge/data_models/api.py:96]: +1

File: src/forge/stores/in_memory_store.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from forge.data_models.api import Store
+
+
+class InMemoryStore(Store):
```
Comment (LucasLLC) [inline src/forge/stores/in_memory_store.py:10]: This is the second PR I see an in memory store implementation, and a bit concerned that this boiler plate code is being generated as a result of torchstore! 

Do you think it'd be a good idea to add a 'single process in memory' version of this to torchstore?

File: src/forge/data_models/api.py (hunk 8 above)
Comment (Ritesh1905) [reply src/forge/data_models/api.py:20]: ack. Will rebase and remove this once the other PR gets merged.

File: src/forge/trainers/huggingface_trainer.py (hunk 1 above)
Comment (LucasLLC) [reply src/forge/trainers/huggingface_trainer.py:18]: Having a trainer abstraction feels like a huge benefit -- users may have lots of different reasons for trying different trainers and they should occupy similar shapes

File: src/forge/generators/vllm_generator.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    def __init__(self, model_path: str):
+        self.model_path = model_path
+
+    def generate(self, prompt: Prompt) -> List[Completion]:
+        """
+        Generate completions for a given prompt using vLLM.
+        """
+        return []
+
+    def update_weights(self, weights_handle: WeightsBuffer) -> None:
```
Comment (pbontrager) [inline src/forge/generators/vllm_generator.py:25]: Currently we have update_weights in Forge as well but I don't think it's compatible with efficient weight updates when the policy is using a continuous batching system. Essentially vLLM decodes a fixed number of tokens per step, it then removes completed requests from the stack and adds new request to fill the stack back up. So the batch is never done. This means that you either have to exhaust the request queue before calling update, which isn't very efficient but gives you full control, or you have to update mid decoding.

Ideally you want to be able to either update as soon as you can (e.g. after the weights have been pre-fetched) or you want want to call generate with a policy request (policy.generate(prompt, policy_version) and then update the policy as soon as all the old policy requests are gone. You would also want to do this at different times for every replica in the service as their queue's would run out at different times.

All of this to say, this logic would be much easier to internalize within the generator. If we put it in the controller, we'll have to have a separate loop that reads a bunch of state information about the generator and calls this update.

File: src/forge/generators/vllm_generator.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+from typing import List
+
+from forge.data_models.api import Generator, WeightsBuffer
+
+from forge.data_models.completion import Completion
+from forge.data_models.prompt import Prompt
+
+
+class VLLMGenerator(Generator):
```
Comment (pbontrager) [inline src/forge/generators/vllm_generator.py:15]: Shouldn't this RFC just define the interfaces? So instead of VLLMGenerator it would just be a Generator abstract class.

File: src/forge/stores/in_memory_store.py (hunk 11 above)
Comment (pbontrager) [inline src/forge/stores/in_memory_store.py:10]: I'm not sure this one is worth adding an abstraction for as torchstore is as fundamental to our stack as monarch. We'd essentially just be updating the interface to match store's interface.

File: src/forge/generators/vllm_generator.py (hunk 12 above)
Comment (pbontrager) [reply src/forge/generators/vllm_generator.py:25]: nit: the input parameter would likely be the policy version or key in store
